zebra: giant zapi cleanup

qlyoung opened the pull request:
#### Standardize arguments for ZAPI message handlers

A lot of the handler functions that are called directly from the ZAPI
input processing code take different argument sets where they don't need
to. These functions are called from only one place and all have the same
fundamental information available to them to do their work. There is no
need to specialize what information is passed to them; it is cleaner and
easier to understand when they all accept the same base set of
information and extract what they need inline.

#### Eliminate the humongous switch-case in the dispatcher

After doing the above work it becomes possible to eliminate the big
switch-case statement and turn it into a lookup table keyed by ZAPI
message code. Granted the compiler usually optimizes switch-case
statements into jump tables anyway but it is better to make this a
strong guarantee by doing it in the source.

#### Allow batching ZAPI I/O

ZAPI's original I/O loop looks like:

`zserv_read`:
- Read one message
- Run message handler
- Schedule `zserv_read`

This has been changed to:

`zserv_handle_messages`:
- For each message on the input queue:
  - Pop the message
  - Run message handler

`zserv_read`:
- Read as many messages as available into input queue
- Schedule `zserv_handle_messages`
- Schedule `zserv_read`

This is more flexible as it decouples I/O from actual state changes inside
Zebra. Theoretically it also helps cache churn by grouping the same tasks
together instead of alternating between them.

#### Modify all handlers to stop accessing global client buffers

There is no reason for message handlers to look at a global data
structure that holds the last message received. Instead we can reduce
the amount of global state they need to know about by making them accept
a message stream, and not care about where it came from. This is also
necessary for the input queue changes above. It also moves us closer to
being able to unit test ZAPI message handlers.

#### Modify all handlers to not return an status code

All of the ZAPI message handlers return an integer that means different
things to each of them, but nobody ever reads these integers, so this is
technical debt that we can just eliminate outright.

#### Rename some functions to `zread_*` or `zsend_*` to clarify action

zserv.c functions are roughly organized into three kinds of functions:
- Functions that encode Zebra datastructures into ZAPI message
  equivalents
- Functions that build and send whole messages
- Functions that process received ZAPI messages

These are variously named with prefixes like "zserv", "zebra", "zread",
"zsend", etc. Organize them by using consistent naming prefixes;
`zserv_encode_*`, `zsend_*`, `zread_*` respectively for the three types given
above.

#### Organize `zserv.c` into logical sections

Group the above functions together by type, and group the plumbing for
the I/O code together as well.

#### Add `struct zmsghdr`

Formalize the ZAPI header by documenting it in code and providing it to
message handlers free of charge to reduce complexity.

#### Clean up zserv_read

Just some simplifications of the existing code intended to make it
easier to read, understand and debug. Also improved debugging output for
ZAPI messages by leveraging zmsghdr and making log messages consistent:

```
2018/03/05 16:00:12 OSPF: ospfd 4.1-dev starting: vty@2604
2018/03/05 16:00:12 ZEBRA: zebra message comes from socket [19]
2018/03/05 16:00:12 ZEBRA: Rx'd ZAPI message
2018/03/05 16:00:12 ZEBRA:  Length: 4
2018/03/05 16:00:12 ZEBRA: Command: ZEBRA_HELLO
2018/03/05 16:00:12 ZEBRA:     VRF: 0
2018/03/05 16:00:12 ZEBRA:
0x0000000001773ad0: 00 0e fe 05             ....

2018/03/05 16:00:12 ZEBRA: Read 1 packets
2018/03/05 16:00:12 ZEBRA: client 19 says hello and bids fair to announce only ospf routes
2018/03/05 16:00:12 ZEBRA: zebra message comes from socket [19]
2018/03/05 16:00:12 ZEBRA: Rx'd ZAPI message
2018/03/05 16:00:12 ZEBRA:  Length: 4
2018/03/05 16:00:12 ZEBRA: Command: ZEBRA_BFD_CLIENT_REGISTER
2018/03/05 16:00:12 ZEBRA:     VRF: 0
2018/03/05 16:00:12 ZEBRA:
0x0000000001773ad0: 00 0e fe 05             ....

2018/03/05 16:00:12 ZEBRA: Read 1 packets
2018/03/05 16:00:12 ZEBRA: bfd_client_register msg from client ospf: length=14
2018/03/05 16:00:12 ZEBRA: zebra message comes from socket [19]
2018/03/05 16:00:12 ZEBRA: Rx'd ZAPI message
2018/03/05 16:00:12 ZEBRA:  Length: 0
2018/03/05 16:00:12 ZEBRA: Command: ZEBRA_ROUTER_ID_ADD
2018/03/05 16:00:12 ZEBRA:     VRF: 0
2018/03/05 16:00:12 ZEBRA:

2018/03/05 16:00:12 ZEBRA: Read 1 packets
2018/03/05 16:00:12 ZEBRA: zebra message comes from socket [19]
2018/03/05 16:00:12 ZEBRA: Rx'd ZAPI message
2018/03/05 16:00:12 ZEBRA:  Length: 0
2018/03/05 16:00:12 ZEBRA: Command: ZEBRA_INTERFACE_ADD
2018/03/05 16:00:12 ZEBRA:     VRF: 0
2018/03/05 16:00:12 ZEBRA:
```

Signed-off-by: Quentin Young <qlyoung@cumulusnetworks.com>

qlyoung commented:
Need to leave `stream_failure` labels defined even when they are unused; in this case we have to put a `return` statement at the end of the function. The style checker will warn about this since usually `return` statements at the end of a function are unnecessary, but these warnings can be safely ignored. It will also warn that `ZAPI_HANDLER_ARGS` needs an identifier name, but this is a macro so this warning can also be ignored.

qlyoung commented:
Deleted spurious comment generated by a broken rebase.

donaldsharp commented on zebra/zserv.c:
```diff
@@ -2569,11 +2569,32 @@ static void zread_vrf_label(struct zserv *client,
 	return;
 }
 
+/*
```
This is zclient_read_header?  If we are doing this... let's do this?

donaldsharp commented on zebra/zserv.c:
```diff
@@ -842,14 +848,14 @@ static void zserv_fec_register(struct zserv *client, u_short length)
 	}
 
-	while (l < length) {
+	while (l < hdr->length) {
```
as a question -> the decode function ( that I commented on in the previous commit ) does not take into account subtraction of the zapi header.  Is this going to be a problem here?

qlyoung replied:
No, the handler functions expect the stream pointer to be at the first byte after the header.
`zserv_process_messages` resets the stream pointer and then extracts the header before calling down to the handlers, so they get the offset they expect.

qlyoung commented:
Rebase

rwestphal commented:
Great work. While reviewing this, I found out that the NetDEF CI is stuck in an infinite loop inside one topotest (`test_ldp_vpls_topo1.py`). This is happening because zebra is crashing when it is configured with the `debug zebra packet` command. Here's the backtrace:
```
#0  0x00007ffff7b5bd61 in zlog_hexdump (mem=0x3535353535303030, len=813182986) at lib/log.c:1103
#1  0x00005555555c9460 in zserv_log_message (errmsg=0x0, msg=0x555555956150, hdr=0x7fffffffe5a4) at zebra/zserv.c:2585
#2  0x00005555555c9cbf in zserv_read (thread=0x7fffffffe770) at zebra/zserv.c:2846
#3  0x00007ffff7b88702 in thread_call (thread=0x7fffffffe770) at lib/thread.c:1497
#4  0x00007ffff7b58258 in frr_run (master=0x5555558085a0) at lib/libfrr.c:879
#5  0x000055555557d4dd in main (argc=1, argv=0x7fffffffea28) at zebra/main.c:377
```

rwestphal commented:
The problem seems to be in the `zlog_hexdump()` function itself, which was not touched by this PR. The diff below fixed the crash for me:
```diff
@@ -1089,7 +1089,7 @@ void zlog_hexdump(const void *mem, unsigned int len)
        unsigned long i = 0;
        unsigned int j = 0;
        unsigned int columns = 8;
-       char buf[(len * 4) + ((len / 4) * 20) + 30];
+       char buf[BUFSIZ];
        char *s = buf;
 
        for (i = 0; i < len + ((len % columns) ? (columns - len % columns) : 0);
```

As odd as it looks, variable length arrays are valid in C99. Changing the size of the array to `(len * 4) + ((len * 20) / 4) + 30` also fixed the crash, apparently the problem was the truncation from `len / 4`.

However my LDP test topology is still not working because zebra is somehow droping the routes received from ospfd and ospf6d. I'll keep investigating.

rwestphal commented on lib/zclient.c:
```diff
@@ -322,6 +322,18 @@ int zclient_read_header(struct stream *s, int sock, u_int16_t *size,
+	STREAM_GETW(zmsg, hdr->length);
+	STREAM_GETC(zmsg, hdr->marker);
+	STREAM_GETC(zmsg, hdr->version);
+	STREAM_GETC(zmsg, hdr->vrf_id);
```
This needs to be `STREAM_GETL()`.

rwestphal commented on zebra/zserv.c:
```diff
@@ -2754,157 +2691,206 @@ static void zserv_write_incoming(struct stream *orig, uint16_t command)
+		/* process commands */
+		zserv_handle_commands(client, &hdr, msg, zvrf);
+
+	} while (msg);
```
In `zebra_client_read()` we read up to `zebrad.packets_to_process` packets at time. Wouldn't it be a good idea to have a limit like that here as well? I think it will be necessary especially when the ZAPI I/O is segregated into another thread reading packets non-stop. And instead of `zebrad.packets_to_process` we could try using `thread_should_yield()` but I'm not sure if it would work.

qlyoung replied:
You are correct, I have code for this on my multithreaded ZAPI branch. I can pull it into this PR if you think it belongs here.

rwestphal replied:
Ok, adding a limit later together with the MT ZAPI code sounds reasonable to me.

qlyoung commented on zebra/label_manager.c:
```diff
@@ -87,7 +89,7 @@ static int relay_response_back(struct zserv *zserv)
 	/* send response back */
 	stream_copy(dst, src);
-	ret = writen(zserv->sock, dst->data, stream_get_endp(dst));
+	ret = writen(zserv->sock, src->data, stream_get_endp(src));
```
It was at a previous version of this PR, not anymore. As I read it there's no difference either way. I can back it out if you like.

qlyoung commented on zebra/zserv.c:
```diff
@@ -2754,157 +2691,206 @@ static void zserv_write_incoming(struct stream *orig, uint16_t command)
+			const char *emsg = "Message has corrupt header";
+			zserv_log_message(emsg, msg, NULL);
+		}
+		if (!hdrvalid)
```
Yeah I tried it the second way and it really crams it in there to fit under 80 characters, I would like to simplify this code some more so there's less indentation but I thought it was more readable the way I have it. I agree it's less than ideal and our ideas of 'readable' differ, I don't care too much either way and will change it back.

qlyoung commented on zebra/zserv.c:
```diff
@@ -2754,157 +2691,206 @@ static void zserv_write_incoming(struct stream *orig, uint16_t command)
+	struct zmsghdr hdr;
+	ssize_t nb;
+	bool hdrvalid;
+	char errmsg[256];
```
fair enough

qlyoung commented:
@rwestphal Fixed a couple of the nits you mentioned. In `zlog_hexdump` I opted to use the order of operations that's not prone to truncation error instead of `BUFSIZ`, since it's plausible that someone will want to hexdump more data than `BUFSIZ` can hold at some point in the future. I also changed all the `sprintf`'s in that function to `snprintf`'s just to be safe. Did the `STREAM_GETC` --> `STREAM_GETL` change and rebased on master. Should be g2g now.
